src/server/chat.js

```diff
@@ -0,0 +1,292 @@
+const { mojangPublicKeyPem } = require('./constants')
+const crypto = require('crypto')
+const { concat } = require('../transforms/binaryStream')
+
+const lastSeenStates = {
+  REMOVED_MESSAGES: 0,
+  OUT_OF_ORDER: 1,
+  UNKNOWN_MESSAGES: 2,
+  DUPLICATED_PROFILES: 3
+}
+
+const chatVisibility = {
+  FULL: 0,
+  SYSTEM: 1,
+  HIDDEN: 2
+}
+
+module.exports = function (client, server, options) {
+  const mojangPubKey = crypto.createPublicKey(mojangPublicKeyPem)
+  const mcData = require('minecraft-data')(client.version)
+  const raise = (translatableError) => client.end(translatableError, JSON.stringify({ translate: translatableError }))
+  const throwGeneric = (message) => client.end(message, JSON.stringify({ translate: 'disconnect.genericReason', with: [{ text: message }] }))
+
+  const {
+    'enable-chat-signing': signedChat
+  } = options
+
+  client.on('chat_message', handleChatMessage)
+
+  client.on('settings', handleClientSettings)
+
+  client.on('player_chat', handleIncomingChat) // Do server-side clients emit this event?
+
+  client.on('message_acknowledgement', handleAcknowledgements)
+
+  client.on('chat_session', handleChatSession)
+
+  client.lastMessage = {
+    timestamp: 0
+  }
+
+  client.packAcknowledgements = (acknowledgements) => {
+    if (!client.signature_cache) client.signature_cache = []
+
+    return acknowledgements.map(ack => {
+      const idx = client.signature_cache.findIndex(el => el.compare(ack) === 0)
+      if (idx === -1) return { id: 0, signature: ack }
+      return { id: idx }
+    })
+  }
+
+  function handleChatSession (packet) {
+    client.session = {
+      uuid: packet.sessionUUID,
+      index: 0
+    }
+
+    const publicKey = crypto.createPublicKey({ key: packet.publicKey, format: 'der', type: 'spki' })
+    const publicDER = publicKey.export({ type: 'spki', format: 'der' })
+    const publicPEM = mcPubKeyToPem(publicDER)
+
+    if (!crypto.verify('RSA-SHA1', concat('UUID', client.uuid, 'i64', packet.expireTime, 'buffer', publicDER), mojangPubKey, packet.signature)) {
+      raise('multiplayer.disconnect.invalid_public_key_signature')
+      return
+    }
+
+    if (!client.profileKeys) {
+      client.verifyMessage = (packet, acknowledgements) => {
+        const length = Buffer.byteLength(packet.message, 'utf8')
+        const previousMessages = acknowledgements.length > 0 ? ['i32', acknowledgements.length, 'buffer', Buffer.concat(...acknowledgements)] : ['i32', 0]
+
+        const signable = concat('i32', 1, 'UUID', client.uuid, 'UUID', client.session.uuid, 'i32', client.session.index, 'i64', packet.salt, 'i64', packet.timestamp / 1000n, 'i32', length, 'pstring', packet.message, ...previousMessages)
```

Comment:
I see this kind of packaging at many places
is it really different every time ?

Reply:
yeah, and this will not work for 1.19.2. Not sure what the `...previousMessages` part is for but it may be a 1.19.3 thing I've not looked into

Reply:
This is for the 1.19.3 chat session packet. The function in login.js is for the 1.19/1.19.2 login start packet. It is needed in both places.